tracetools/span.go

```diff
@@ -2,28 +2,105 @@ package tracetools
 
 import (
 	"context"
+	"fmt"
 
 	"github.com/opentracing/opentracing-go"
 	"github.com/opentracing/opentracing-go/ext"
-	"github.com/opentracing/opentracing-go/log"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	"go.opentelemetry.io/otel/trace"
 	ddext "gopkg.in/DataDog/dd-trace-go.v1/ddtrace/ext"
 )
 
-// StartSpanFromContext will start a span from the given context with the given
-// operation name. It will also do some common/repeated setup on the span to keep
-// code a little more DRY.
-func StartSpanFromContext(ctx context.Context, operation string) (opentracing.Span, context.Context) {
-	span, ctx := opentracing.StartSpanFromContext(ctx, operation)
-	// Make the span available for analytics in Datadog
-	span.SetTag(ddext.AnalyticsEvent, true)
-	return span, ctx
+// StartSpanFromContext will start a span appropriate to the given tracing backend from the given context with the given
+// operation name. It will also do some common/repeated setup on the span to keep code a little more DRY.
+// Will panic if tracerType is anything other than "datadog", "otel", or ""
+func StartSpanFromContext(ctx context.Context, operation string, tracerType string) (any, context.Context) {
+	switch tracerType {
+	case "datadog":
+		span, ctx := opentracing.StartSpanFromContext(ctx, operation)
+		span.SetTag(ddext.AnalyticsEvent, true) // Make the span available for analytics in Datadog
+		return span, ctx
+
+	case "otel":
+		ctx, span := otel.Tracer("buildkite_agent").Start(ctx, operation)
+		span.SetAttributes(attribute.String("analytics.event", "true"))
+		return span, ctx
+
+	case "":
+		return nil, ctx
+
+	default:
+		panic(fmt.Sprintf("Invalid tracing backend %q passed to tracetools.StartSpanFromContext", tracerType))
```

Comment:
i'm not sure that panicking is the right call here - it means that we crash if we get an invalid backend, rather than just not tracing. maybe silently dropping traces is bad too though?